### PR TITLE
GH-920: Propagate afterReceive to AsyncRTemplate

### PR DESCRIPTION
**Pull request by artembilan**
Fixes https://github.com/spring-projects/spring-amqp/issues/920

The `AsyncRabbitTemplate` is missing an after receive post-processing
in its internal containers for replies

* Reuse `afterPostProcessors` from the provided `RabbitTemplate` and
propagate them into the internal listener containers for replies
* Fix `AbstractCompressingPostProcessor` do not mutate provided
`MessageProperties` and build a fresh instance.
* Demonstrate post-processors propagation by the `GZipPostProcessor`
& `GUnzipPostProcessor` configuration in the `AsyncRabbitTemplateTests`

**Cherry-pick to 2.1.x**

### CODE REVIEW

**Comment by artembilan** [comment]
This is not going to pass on Travis though, because of https://github.com/spring-projects/spring-framework/issues/22702

**Comment by philwebb** [comment]
@artembilan I just pushed a fix for https://github.com/spring-projects/spring-framework/issues/22702

**Comment by artembilan** [comment]
Fully not sure why it still fails on Travis.
Everything is OK locally. 
Please, check on merge.
Thanks

File: spring-amqp/src/main/java/org/springframework/amqp/support/postprocessor/AbstractCompressingPostProcessor.java [hunk 1]
```diff
@@ -74,13 +78,19 @@ public Message postProcessMessage(Message message) throws AmqpException {
 		try {
 			OutputStream zipper = getCompressorStream(zipped);
 			FileCopyUtils.copy(new ByteArrayInputStream(message.getBody()), zipper);
-			MessageProperties messageProperties = message.getMessageProperties();
-			String currentEncoding = messageProperties.getContentEncoding();
-			messageProperties
-					.setContentEncoding(getEncoding() + (currentEncoding == null ? "" : ":" + currentEncoding));
+			MessageProperties originalProperties = message.getMessageProperties();
+			MessageBuilderSupport<MessageProperties> messagePropertiesMessageBuilder =
```
Comment (garyrussell) [inline spring-amqp/src/main/java/org/springframework/amqp/support/postprocessor/AbstractCompressingPostProcessor.java]: Why the abstract type here instead of `MessagePropertiesBuilder` ?

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java [hunk 2]
```diff
@@ -157,6 +158,11 @@ public AsyncRabbitTemplate(ConnectionFactory connectionFactory, String exchange,
 		this.template.setExchange(exchange == null ? "" : exchange);
 		this.template.setRoutingKey(routingKey);
 		this.container = new SimpleMessageListenerContainer(connectionFactory);
+		Collection<MessagePostProcessor> afterReceivePostProcessors = this.template.getAfterReceivePostProcessors();
+		if (afterReceivePostProcessors != null) {
+			this.container.setAfterReceivePostProcessors(
```
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: If not null? `JavaUtils` ?

File: spring-amqp/src/main/java/org/springframework/amqp/support/postprocessor/AbstractCompressingPostProcessor.java [hunk 3]
```diff
@@ -74,13 +78,19 @@ public Message postProcessMessage(Message message) throws AmqpException {
...
+					MessagePropertiesBuilder.fromClonedProperties(originalProperties)
```
Comment (garyrussell) [inline spring-amqp/src/main/java/org/springframework/amqp/support/postprocessor/AbstractCompressingPostProcessor.java]: It's not at all clear why this change is needed. Seems like unnecessary overhead.

File: spring-amqp/src/main/java/org/springframework/amqp/support/postprocessor/AbstractCompressingPostProcessor.java (hunk 1 above)
Comment (artembilan) [reply spring-amqp/src/main/java/org/springframework/amqp/support/postprocessor/AbstractCompressingPostProcessor.java]: That is what returned by the `setContentEncoding()`.

I guess I can rework the logic here into the `MessageBuilder` instead...

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java (hunk 2 above)
Comment (artembilan) [reply spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: Good point!

File: spring-amqp/src/main/java/org/springframework/amqp/support/postprocessor/AbstractCompressingPostProcessor.java (hunk 3 above)
Comment (artembilan) [reply spring-amqp/src/main/java/org/springframework/amqp/support/postprocessor/AbstractCompressingPostProcessor.java]: We build here a new message and modify existing properties new new content encoding like `gzip:`.
So, the original message gets wrong properties. 
We don't here on the producer side a guarantee that the provided message is not built by end-user.
In fact the `AsyncRabbitTemplateTests` uses the same `fooMessage` for different sends and I only figured out in debug that it fails because of `gzip:gzip:` prefix in the content encoding for the second send.
It is OK on the consumer side though, since message is created by the Framework and we really don't propagate it to the target end-user code.

Does it make sense?

Comment (garyrussell) [reply spring-amqp/src/main/java/org/springframework/amqp/support/postprocessor/AbstractCompressingPostProcessor.java]: Got it; thanks, but perhaps it should be an option (default dont clone,)? In most cases, users won't keep the original message around - especially if `convertAndSend()` is used.

**Comment by garyrussell** [comment]
Merged as 38358d306b04c8df47da44970669933e473b5856 with minor doc polishing, and cherry-picked to 2.1.x as 142fd7f44322703b9a932d2fe285413717c974a9
